### PR TITLE
Implementing VectorSumCombiner

### PR DESCRIPTION
**Pull request by rialg**
## Description
Following the first item described in #264, this PR has the implementation of the `VectorSumCombiner` class, which should be used to compute vector/array sums.

## Affected Dependencies
Needed to extend [AggregateParams](https://github.com/OpenMined/PipelineDP/blob/3a7b87a5d943ea6b083c45934c8bb1110cbe5766/pipeline_dp/aggregate_params.py#L56) with the fields `norm_kind` and `max_norm`.

## How has this been tested?
* Tests added to `combiners_test.py`
```
==================================================================================================================== test session starts =====================================================================================================================
platform linux -- Python 3.9.10, pytest-7.1.2, pluggy-1.0.0
plugins: timeout-2.1.0
collected 45 items                                                                                                                                                                                                                                           

tests/combiners_test.py .............................................                                                                                                                                                                                  [100%]

====================================================================================================================== warnings summary ======================================================================================================================
../.local/lib/python3.9/site-packages/hdfs/config.py:15
DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    from imp import load_source

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================================================================== 45 passed, 1 warning in 2.20s ================================================================================================================
```

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pipeline_dp/combiners.py [hunk 1]
```diff
@@ -148,6 +149,18 @@ def mean_var_params(self):
             self.aggregate_params.max_contributions_per_partition,
             self.aggregate_params.noise_kind)
 
+    @property
+    def additive_vector_noise_params(self):
+        return dp_computations.AdditiveVectorNoiseParams(
+            eps_per_coordinate=self.eps,
```
Comment (dvadym) [inline pipeline_dp/combiners.py:155]: eps_per_coordinate = eps/array_size
delta_per_coordinate = delta/array_size

File: examples/movie_view_ratings/run_all_frameworks.py [hunk 2]
```diff
@@ -79,12 +80,22 @@ def calc_dp_rating_metrics(movie_views, backend, public_partitions):
         min_value=1,
         max_value=5)
 
+    value_extractor = lambda mv: mv.rating
+
+    if FLAGS.vector_metrics:
```
Comment (rialg) [inline examples/movie_view_ratings/run_all_frameworks.py:85]: @dvadym , I've added a flag to compute vector metrics over the ratings values, which are extracted as vectors with zeros and one in the position of the rating. What other example would be useful to add?

File: pipeline_dp/combiners.py [hunk 3]
```diff
@@ -507,6 +559,17 @@ def create_compound_combiner(
         combiners.append(
             PrivacyIdCountCombiner(
                 CombinerParams(budget_privacy_id_count, aggregate_params)))
+    if pipeline_dp.Metrics.ARRAY_SUM in aggregate_params.metrics:
+        budget_vector_sum = budget_accountant.request_budget(
+            mechanism_type, weight=aggregate_params.budget_weight)
+        combiners.append(
+            VectorSumCombiner(
+                CombinerParams(budget_vector_sum, aggregate_params)))
+        if pipeline_dp.Metrics.SUM in aggregate_params.metrics or \
```
Comment (dvadym) [inline pipeline_dp/combiners.py:568]: I thought again, where to check, we should fail eariler, in this case in `[AggregateParams.__post_init()](https://github.com/OpenMined/PipelineDP/blob/3a7b87a5d943ea6b083c45934c8bb1110cbe5766/pipeline_dp/aggregate_params.py#L86)`. Could you please move it there?

File: pipeline_dp/combiners.py [hunk 4]
```diff
@@ -507,6 +559,17 @@ def create_compound_combiner(
...
+            pipeline_dp.Metrics.MEAN in aggregate_params.metrics or \
+            pipeline_dp.Metrics.COUNT in aggregate_params.metrics or \
+            pipeline_dp.Metrics.VARIANCE in aggregate_params.metrics:
+            raise ValueError("Do not combine scalar with vector metrics")
```
Comment (dvadym) [inline pipeline_dp/combiners.py:572]: "Vector sum can not be computed together with scalar metrics, like sum, mean etc"

File: pipeline_dp/combiners.py [hunk 5]
```diff
@@ -507,6 +559,17 @@ def create_compound_combiner(
...
                 CombinerParams(budget_privacy_id_count, aggregate_params)))
+    if pipeline_dp.Metrics.ARRAY_SUM in aggregate_params.metrics:
+        budget_vector_sum = budget_accountant.request_budget(
+            mechanism_type, weight=aggregate_params.budget_weight)
+        combiners.append(
+            VectorSumCombiner(
+                CombinerParams(budget_vector_sum, aggregate_params)))
+        if pipeline_dp.Metrics.SUM in aggregate_params.metrics or \
+            pipeline_dp.Metrics.MEAN in aggregate_params.metrics or \
+            pipeline_dp.Metrics.COUNT in aggregate_params.metrics or \
```
Comment (dvadym) [inline pipeline_dp/combiners.py:570]: Please remove COUNT from this list (count can be computed)

File: pipeline_dp/combiners.py [hunk 6]
```diff
@@ -20,6 +20,7 @@
 from pipeline_dp import budget_accounting
 import numpy as np
 import collections
+from pipeline_dp.aggregate_params import NormKind
```
Comment (dvadym) [inline pipeline_dp/combiners.py:23]: According to Google styleguide, it's needed to import modules, not classes. Could you please
1.Add NormKind to [__init__.py](https://github.com/OpenMined/PipelineDP/blob/main/pipeline_dp/__init__.py) (there are public API objects).
2. Use `pipelinedp.NormKind` in the code

**Review by dvadym — CHANGES_REQUESTED** [review]
Thank you! It looks good, I left comments

File: pipeline_dp/combiners.py [hunk 7]
```diff
@@ -460,6 +475,48 @@ def metrics_names(self) -> List[str]:
...
+    AccumulatorType = np.ndarray
+
+    def __init__(self, params: CombinerParams):
+        self._params = params
+
+    def create_accumulator(self,
+                           values: Iterable[ArrayLike]) -> AccumulatorType:
+        array_sum = None
+        for val in values:
+            if type(val) is not np.ndarray:
```
Comment (dvadym) [inline pipeline_dp/combiners.py:493]: Nit (just learned :) ): More idiomatic way: 
`if not isinstance(val, np.ndarray):`

File: pipeline_dp/aggregate_params.py [hunk 8]
```diff
@@ -69,13 +70,15 @@ class AggregateParams:
     max_value: Upper bound on each value.
     custom_combiners: Warning: experimental@ Combiners for computing custom
       metrics.
+    vector_norm_kind: The type of norm to use for the DP calculations.
```
Comment (dvadym) [inline pipeline_dp/aggregate_params.py:73]: Nit: VECTOR_SUM metric caluation.

File: pipeline_dp/aggregate_params.py [hunk 9]
```diff
@@ -69,13 +70,15 @@ class AggregateParams:
...
+    vector_max_norm: Bound on each value of a vector.
```
Comment (dvadym) [inline pipeline_dp/aggregate_params.py:74]: Please add: Used only for VECTOR_SUM metric.

File: pipeline_dp/aggregate_params.py [hunk 10]
```diff
@@ -69,13 +70,15 @@ class AggregateParams:
...
+    vector_size: Number of coordinates in a vector.
```
Comment (dvadym) [inline pipeline_dp/aggregate_params.py:75]: Please add: Used only for VECTOR_SUM metric.

**Review by dvadym — COMMENTED** [review]
Thank you! It looks great. Just a few minor comments

**Review by dvadym — APPROVED** [review]
Thank you!

**Pull request merged**